In [1]:
import pandas as pd
import numpy as np
data_gdynia = pd.read_csv(r"C:\Users\Dell\Desktop\Zabawa_Gdynia\Wstepny_Csv_Gdynia\Dataset_Gdynia_Śródmieście.csv", sep=";", index_col = 0)
data_gdynia.head()

,osm_id,addr_street,addr_number,grupa_parter,funkcja_parter,grupa_pietra,funkcja_pietra,liczba_kondygnacji,mieszana,rok_budowy_szac,styl_arch,stan_tech,wysokosc_max,zabytek
id,,,,,,,,,,,,,,
1,127184291,Świętego Piotra,8,NaN,Hotel,NaN,Hotel,7.0,NaN,NaN,NaN,NaN,28.8,NaN
2,890729731,Świętego Piotra,6D,NaN,Mieszkaniowa Wielorodzinna,NaN,Mieszkaniowa Wielorodzinna,5.0,NaN,NaN,NaN,NaN,23.8,NaN
3,890292040,Węglowa,22,NaN,Centrum Handlowe,NaN,Mieszkaniowa Wielorodzinna,17.0,NaN,NaN,NaN,NaN,58.5,NaN
4,890729732,Świętego Piotra,6,NaN,Mieszkaniowa Wielorodzinna,NaN,Mieszkaniowa Wielorodzinna,17.0,NaN,NaN,NaN,NaN,59.4,NaN
5,136885205,Tadeusza Wendy,15,NaN,Centrum Medyczne,NaN,Centrum Medyczne,6.0,NaN,NaN,NaN,NaN,31.1,NaN


In [2]:
wys = np.array(data_gdynia.wysokosc_max)
print(type(wys))
print(wys.dtype)
print(wys.shape)
print(wys.ndim)

<class 'numpy.ndarray'>
float64
(1241,)
1


In [3]:
wys_cm_noround = (wys*100).astype(int)
wys_cm_round = np.round(wys*100).astype(int)
diff = wys_cm_noround - wys_cm_round
values = np.unique(diff, return_index = True, return_counts= True)
print(f"wys 1 met: {wys_cm_noround[37]}, wys 2 met: {wys_cm_round[37]}")
print(wys[37]*100)

wys 1 met: 509, wys 2 met: 510
509.99999999999994


astype(int) drops the decimal part without rounding.
the computer stores the value like a little smaller one ---> 5.1 is stored as 5.09999999999
after multiplying by 100, the value is [509.99999999999994]
saving as int saves the number as 509, where saving as int after rounding saves the number as 510

it applies to 105 out of 1241 records

Always round the numbers when converting to astype(int)

In [4]:
kond_heigh = np.array(data_gdynia.wysokosc_max / data_gdynia.liczba_kondygnacji)
print(kond_heigh)
print(np.isnan(kond_heigh).sum())


[4.11428571 4.76       3.44117647 ...        nan        nan        nan]
450


liczba_kondygnacji is float because it contains NaN, and NaN exists only for floats.
int doesnt have a way to save NaN as an integer, so it replaced the NaN values as the smallest stored value in int64 as -9223372036854775808

Out of 1241 records 450 numbers were converted to extraordinarily small number

This can make counting average value [possible], but the value will be wrong,
while using a column with NaN values in the future dont use int or use ["Int64"], from Pandas that also stores NaN

In [5]:
kond_heigh_est = np.array(np.floor(data_gdynia.wysokosc_max/3))
kond_heigh[614] = np.nan
print( np.nanmean(np.abs(kond_heigh - kond_heigh_est)))
print( np.nanmean((kond_heigh - kond_heigh_est)))


6.3564220333526
1.345476700824616


[Dividing wysokosc_max by liczba_kondygnacji produced NaN and inf values.]
Many rows in liczba_kondygnacji are NaN values and during divisions on NaN it returns NaN. During dividing by 0, it returns inf.
[The inf comes from record 615 (Świętojańska 18), which has 0 floors, a data entry error.]

The total of the mentioned outputs are 450 records of NaN value and 1 record of inf values.

In the future first of all check if the dividing column has NaN or 0 values.

In [6]:
data_gdynia.head(10)

,osm_id,addr_street,addr_number,grupa_parter,funkcja_parter,grupa_pietra,funkcja_pietra,liczba_kondygnacji,mieszana,rok_budowy_szac,styl_arch,stan_tech,wysokosc_max,zabytek
id,,,,,,,,,,,,,,
1,127184291,Świętego Piotra,8,NaN,Hotel,NaN,Hotel,7.0,NaN,NaN,NaN,NaN,28.8,NaN
2,890729731,Świętego Piotra,6D,NaN,Mieszkaniowa Wielorodzinna,NaN,Mieszkaniowa Wielorodzinna,5.0,NaN,NaN,NaN,NaN,23.8,NaN
3,890292040,Węglowa,22,NaN,Centrum Handlowe,NaN,Mieszkaniowa Wielorodzinna,17.0,NaN,NaN,NaN,NaN,58.5,NaN
4,890729732,Świętego Piotra,6,NaN,Mieszkaniowa Wielorodzinna,NaN,Mieszkaniowa Wielorodzinna,17.0,NaN,NaN,NaN,NaN,59.4,NaN
5,136885205,Tadeusza Wendy,15,NaN,Centrum Medyczne,NaN,Centrum Medyczne,6.0,NaN,NaN,NaN,NaN,31.1,NaN
6,273937994,Tadeusza Wendy,15,NaN,Biurowa,NaN,Biurowa,9.0,NaN,NaN,NaN,NaN,29.8,NaN
7,398139557,Tadeusza Wendy,15,NaN,Biurowa,NaN,Biurowa,1.0,NaN,NaN,NaN,NaN,27.6,NaN
8,273937986,Tadeusza Wendy,15,NaN,Biurowa,NaN,Biurowa,8.0,NaN,NaN,NaN,NaN,28.9,NaN
9,126117723,Świętego Piotra,19,NaN,Bar,NaN,Mieszkaniowa Wielorodzinna,5.0,NaN,NaN,NaN,NaN,20.6,NaN


In [7]:
print(wys[:10])
print(wys[-10:])
print(wys[::10])
print(wys[::10].shape)

[28.8 23.8 58.5 59.4 31.1 29.8 27.6 28.9 20.6 25.7]
[47.1 44.7 45.  47.  26.6 47.7 47.  47.2 47.1 44.1]
[ 28.8  22.9  23.9   7.9  11.3  22.1  20.4  34.9  28.9  14.9  24.4  23.
  10.9  29.8  24.5  10.4  12.9  24.9  30.6  15.4   7.3  14.3  21.8  21.9
   8.   44.4  17.6 124.   26.2  19.8  22.1  12.2  13.2   9.1   9.6   7.3
  31.6  25.2  25.4  24.7   6.6  29.3   8.1  11.3  43.1   9.2  33.1  11.4
  27.2  23.7  25.4  24.7  30.2  28.4  18.9   7.7  32.3  14.5  28.8   5.7
  23.7  25.5  12.2  14.3  32.6  31.2  30.5  17.2  13.9  13.8  17.9  44.3
  36.5  33.8  28.4  28.7  33.8  33.2  28.2  20.3  19.2  32.6  23.6  21.1
   7.7  30.5  31.5  33.7  33.7  12.8  32.8  33.8  33.2  34.4  34.7  35.4
  17.5  16.2  17.4  40.9  35.6  48.8  26.3  20.1  38.8  35.5  36.9  36.7
  35.4  43.5  40.4  40.3  37.4  21.6  21.7  29.9  39.5  39.8  40.6  40.6
  44.7  39.9  44.4  27.3  44.1]
(125,)


the ::n returns every n element of the table, and to count the number of elements its total number of elements/3(+1 if there is a rest)

In [8]:
#first5 = wys[:5]
#first5[:5] = 0
first5_2 = wys.copy()
first5_2[:5] = 0

#print(first5[:5])
print(wys[:5])
print(first5_2[:5])



[28.8 23.8 58.5 59.4 31.1]
[0. 0. 0. 0. 0.]


declarating the new array out of existing one only changes the etiquete, so while altering the new array it also alters the old one
the copying of the old array only preserve the old valuse from changing. 

for the future copy the values of the table, to not mix the dataset!!!

In [9]:
kond = np.array(data_gdynia.liczba_kondygnacji)

arr2d = np.stack([wys, kond], axis = 1)
print(arr2d.shape)
print(arr2d[99])
print(arr2d[:,1])
print(arr2d.T.shape)

(1241, 2)
[12.1  3. ]
[ 7.  5. 17. ... nan nan nan]
(2, 1241)


np. stack blocks the arrays into higher dimention array based on the chosen axis
.T can transpond the shape the 0,1 => 1,0

In [10]:
rng = np.random.default_rng(42)
print(wys[rng.choice(len(wys), size = 5 , replace = False)])

print(wys[rng.choice(len(wys), size = 5 , replace = False)])


[36.9  7.5 33.7 23.  25.4]
[48.8 25.5 33.9 34.5 34.8]


the generator generates pseudo random numbers based on the seed between the parentheses
so every time the same seed is run, no matter the computer, time or ide the score is the same
when its run the second time the numbers are different than the first the second row also applies to the next values

In [11]:
print(wys.argsort(descending=True)[:5])
data_gdynia.osm_id.iloc[wys.argsort(descending=True)[:5]]

[270 792   3   2 398]


id
271     76208008
793    260675986
4      890729732
3      890292040
399    355431246
Name: osm_id, dtype: int64

wys.argsort creates a list of positions of the highest builings from the series like in the example [270 792   3   2 398]
the numbers are not the same as id, becouse series starts from 0, bnuildings id starts from 1, 

to ensure that the data will be right all the time, the iloc version that connects the data with the values filtered from the top values that will always be right, 
without any checking positions with []
